Refactor dependencies

aedart opened the pull request:
PR refactors the root `CMakeLists.txt` and `tests/CMakeLists.txt` file. All dependencies and dev-dependencies are now extracted into separate *.cmake files.

The newly added `dependencies.cmake` and `dev-dependencies.cmake` rely on the project's "options", which determine what dependencies should be installed / enabled or not.

In addition, a new `install-runtime.sh` script has been added in the root of the project. It installs all known runtime dependencies and dev-dependencies for the project. This will benefit us, once we decide to enable automatic CI tests.

Lastly, the `README.md` has been slightly changed. It now references the new  `install-runtime.sh` script.

## Details

Please see additional details in `CHANGELOG.md`.

aedart committed:
Fix unable to build due to strict compile options

Forgot about the "switch-default" option. This ensures that the project can be built.

aedart committed:
Refactor project version, move into custom module

aedart committed:
Add dependencies cmake file (incomplete)

aedart committed:
Add dev-dependencies cmake file (incomplete)

aedart committed:
Add SDL2 dependency script

aedart committed:
Add FreeType dependency script

aedart committed:
Add Magic Enum C++ dependency script

aedart committed:
Fix reference

Ups.... copy & paste mistake!

aedart committed:
Add cURL dependency script

aedart committed:
Add OpenSSL dependency script

aedart committed:
Set include guard to global

aedart committed:
Link to Open SSL

Previously only the Crypto library was needed, for the Core Lib. However, we will also need the Open SSL lib

aedart committed:
Add MBed TLS dependency script

aedart committed:
Deprecate Find MBed TLS module

Replaced by new external/mbed-tls module.

aedart committed:
Add ZLib dependency script

aedart committed:
Add std::thread dependency script

aedart committed:
Add WPA Supplicant dependency script

aedart committed:
Improve description

find_library caches the result and allows user-provided cache entry description.

aedart committed:
Change status msg from "Found xyz" to "Using xyz"

aedart committed:
Change debug msg style

aedart committed:
Add Core Lib's dependencies

aedart committed:
Refactor, extract dependencies

aedart committed:
Add Doctest dependency script

aedart committed:
Fix incorrect installed version shown

aedart committed:
Add Lighttpd dependency script

aedart committed:
Add Core Lib dev-dependencies

aedart committed:
Define the "test binary" before dev-dependencies included

In this case, this is needed or some of the dependencies are not able to include, link or customise the build against the desired "target".

aedart committed:
Refactor, extract dev-dependencies

aedart committed:
Fix style

aedart committed:
Change release notes

aedart committed:
Add install runtime script

aedart committed:
Chang prerequisites section

Now referencing the new "install-runtime.sh" script.
Also, reformatted the file a bit...

aedart committed:
Add JSON Lint as dev-dependency

This fixes the JSON unit tests - they are now able to pass.

aedart committed:
Change release notes

aedart committed:
Move "fixed" up

kernelguy commented on install-runtime.sh:
Please rename this script to reflect it is build environment dev dependencies

kernelguy replied:
Suggestion: install-dev-libraries.sh

aedart replied:
Done - filename changed.

kernelguy commented on cmake/external/wpa-supplicant.cmake:
```diff
@@ -0,0 +1,33 @@
+
+# find_package(WPA_SUPPLICANT REQUIRED) # NOT going to work without a custom `FindWPASupplicant` module
+
+find_library(wpa_supplicant NAMES "wpa_client" REQUIRED DOC " RSP Core Lib - Path to WPA Supplicant Library (libwpa-client-dev)")
```
This should not be marked as a required dependency.

aedart replied:
Has now been changed to no longer be a required library. However, should this perhaps not be defined as a dev-dependency? We can always move this around at a later point, if such is the case.

aedart committed:
Rename script to "install dev libraries"

As suggested by Steffen, the filename better reflects what it installs.

aedart committed:
Change script name

aedart committed:
(Re)Change WPA Supplicant to longer be a required dependency

Previous commit message stated that this fixed an issue for "GraphQL" - that was incorrect. We are using "CodeQL", which has nothing to do with "GraphQL".

See 52620af70b5c748debfbaa6038458db39d62e085

aedart committed:
Change release notes

aedart committed:
Fix incorrect "not found" check for WPA Supplicant

aedart committed:
Add message concerning the version of CMAKE that is being used

This is useful for when different kinds of CI runs are executed, e.g. CodeQL

aedart committed:
Change dev-dependencies to only be included only when BUILD_TESTING
